A72-open-telemetry-tracing.md

```diff
@@ -0,0 +1,428 @@
+      // Used by other TextMap propagators, e.g. w3c.
+      return carrier.get(Metadata.Key.of(key, ASCII_STRING_MARSHALLER));
+    }
+  }
+
+  // Add a new method to optimize the TextMap propagator to avoid base64 encoding.
+  @Override
+  public byte[] getBinary(@Nullable Metadata carrier, String key) {
+    assert key.endsWith("bin");
+    return carrier.get(Metadata.Key.of(key, BINARY_BYTE_MARSHALLER));
+  }
+}
+
+// This interface will be implemented by gRPCCommonSetter/Getter as an optimization path 
+// to avoid base64 encoding between TextMap APIs due to lack of
+// OpenTelemetry binary propagator API.
+private interface GrpcBinarySetter<Metadata> {
+    void set(Metadata header, String key, byte[] value);
+}
+
+```
+
+The `GrpcCommonSetter` adds an overloaded `set()` method to directly take `bytes[]` 
+(Java and Go) to avoid extra base64 encoding. For the normal `set()` method it 
+should handle both binary (`-bin`) header and ASCII header from any TextMap 
+propagators that users may config.
+The `GrpcCommonGetter` adds new method `getBinary()` for the optimized path for 
+the same reason in Java and Go. Similarly, the normal `get()` method handles both 
+binary header and TextMap propagators.
+
+#### Context Propagation APIs in C++
+C++ will also support the propagators API, because this imposes API 
+uniformity among languages. Due to the language restriction, 
+C++ can not take the optimization path to workaround lacking the binary 
+propagator API. That means using propagators API with C++ needs base64 encoding 
+and therefore is slower compared with just using metadata API. However, C++ can be 
+configured to interact with metadata directly, like the current gRPC OpenCensus.
+This is a faster way that avoids paying for the performance cost due to 
+string/binary encoding between the propagator and the getter/setter. 
+We use this strategy to balance between API simplicity and performance efficiency.
+The two APIs C++ will support for the context propagation are:
+* If `grpcTraceBinPropagator` is configured, take a slower path in the pseudocode 
+described above. 
+* If explicitly configured, gRPC will directly use `Metadata.get()` and `Metadata.put()` 
+APIs on the `grpc-trace-bin` header. No TextMapPropagator API and TextMapSetter/Getter 
+will be involved. This is a faster path and mitigates performance concerns due 
+to base64 encoding.
+
+TODO: add pseudocode here in C++ for `grpcTraceBinPropagator`,`GrpcCommonSetter` 
+and `GrpcCommonSetter`.
+
+The `GrpcCommonSetter.set()` and `GrpcCommonGetter.get()` method in C++
+should handle both binary (`-bin`) header 
+(e.g. `grpc-trace-bin`) and ASCII header from other text map 
+propagators that users configure into gRPC OpenTelemetry, e.g. w3c.
+
+### Grpc OpenTelemetry Tracing API
+This section talks about enabling and configuring OpenTelemetry tracing.
+The OpenTelemetry API will coexist with the 
+OpenCensus API until the latter is dropped. Only one "grpc-trace-bin" header 
```

Comment:
Let's remove "until the latter is dropped." That isn't part of this gRFC, and we don't have a schedule for it yet.